internal/core/metadata/init.go

```diff
@@ -108,38 +118,90 @@ func initializeConfiguration(useConsul bool, useProfile string) (*ConfigurationS
 }
 
 func connectToConsul(conf *ConfigurationStruct) error {
-	// Initialize service on Consul
-	err := consulclient.ConsulInit(consulclient.ConsulConfig{
-		ServiceName:    internal.CoreMetaDataServiceKey,
-		ServicePort:    conf.ServicePort,
-		ServiceAddress: conf.ServiceAddress,
-		CheckAddress:   conf.ConsulCheckAddress,
-		CheckInterval:  conf.CheckInterval,
-		ConsulAddress:  conf.ConsulHost,
-		ConsulPort:     conf.ConsulPort,
-	})
+	//Obtain ConsulConfig
+	cfg := consulclient.NewConsulConfig(conf.Registry, conf.Service, internal.CoreMetaDataServiceKey)
+	// Register the service in Consul
+	err := consulclient.ConsulInit(cfg)
+
 	if err != nil {
 		return fmt.Errorf("connection to Consul could not be made: %v", err.Error())
 	} else {
 		// Update configuration data from Consul
-		if err := consulclient.CheckKeyValuePairs(conf, internal.CoreMetaDataServiceKey, strings.Split(conf.ConsulProfilesActive, ";")); err != nil {
-			return fmt.Errorf("error getting key/values from Consul: %v", err.Error())
+		updateCh := make(chan interface{})
+		errCh := make(chan error)
+		dec := consulclient.NewConsulDecoder(conf.Registry)
+		dec.Target = &ConfigurationStruct{}
+		dec.Prefix = internal.ConfigV2Stem + internal.CoreMetaDataServiceKey
+		dec.ErrCh = errCh
+		dec.UpdateCh = updateCh
+
+		defer dec.Close()
+		defer close(updateCh)
+		defer close(errCh)
+		go dec.Run()
+
+		select {
+		case <-time.After(2 * time.Second):
+			err = errors.New("timeout loading config from registry")
+		case ex := <-errCh:
+			err = errors.New(ex.Error())
+		case raw := <-updateCh:
+			actual, ok := raw.(*ConfigurationStruct)
+			if !ok {
+				return errors.New("type check failed")
+			}
+			Configuration = actual
+		}
+
+		if err != nil {
+			return err
 		}
 	}
 	return nil
 }
 
+func listenForConfigChanges() {
+	errCh := make(chan error)
+	dec := consulclient.NewConsulDecoder(Configuration.Registry)
+	dec.Target = &ConfigurationStruct{}
+	dec.Prefix = internal.ConfigV2Stem + internal.CoreMetaDataServiceKey
+	dec.ErrCh = errCh
+	dec.UpdateCh = chConfig
+
+	defer dec.Close()
+	defer close(errCh)
+
+	go dec.Run()
+	for {
+		select {
+		case ex := <-errCh:
+			LoggingClient.Error(ex.Error())
+		case raw, ok := <-chConfig:
+			if ok {
+				actual, ok := raw.(*ConfigurationStruct)
+				if !ok {
+					LoggingClient.Error("listenForConfigChanges() type check failed")
+				}
+				Configuration = actual //Mutex needed?
```

Comment:
Should we reconnect the db and clients here? Some of the services ip(db, logging...) could be changed.

Reply:
@feclare Well, the service clients already have the Endpointer.Monitor() loop which looks for updated Consul information on a configurable interval. But you're right that we could look at making those endpoints refresh as part of the above flow. And also the database client doesn't have a Monitor to refresh itself if the connection params change. Might be good work for an enhancement PR.